client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+Specifically, every entry in a snapshot of the Treode database has both read and value timestamps.  A snapshot is consistent if and only if (maximum value timestamp) < (minimum read timestamp).  (See [this explanation](https://forum.treode.com/t/eventual-consistency-and-transactions-working-together/36).)
+
+Then, by the invariant established previously that each entry is guaranteed to be the value in the DB between [V\_t, R\_t], it follows that there was a point in time when the DB contained all the values in the snapshot simultaneously.
+
+The Transaction uses min\_Rt and max\_Vt to maintain this invariant and avoid wasting network time sending commits to the DB server guaranteed to be rejected.  If any Transaction read causes (max\_Vt > min\_Rt), the read method throws a StaleException(Vt), including the value time that caused the violation.  (Note that a read time should not cause a violation because we can simply set min-age=0. Contrarily, we cannot control the value times through our read query parameters.) 
+
+### min\_Rt 
+
+The Transaction maintains the minimum read time it has seen so far.  On every read, if the read time is less than min\_Rt, we update min\_Rt.  
+
+Note that this read time is NOT the time the Transaction read the entry.  Transaction reads from cache do not update entry read times; cache reads from the DB server do update entry read times. 
+
+### max\_Vt
+
+The Transaction maintains the maximum value time it has seen so far.  On every read, if the value time is greater than max\_Vt, we update max\_Vt.
+
+Note that updating max\_Vt can cause the invariant violation described above, causing the Transaction read method to throw StaleException(Vt).  Intuitively, when reading an entry from the DB, we have no idea when it was last updated until we actually read it.  As a result, we cannot avoid the possibility of violating our consistent snapshot invariant.
+
+* **Question (6):** If the user fails to specify max\_age correctly, should we throw a StaleException?
+
+### view 
+
+The view maintains the set of operations performed throughout the Transaction.  When the user commits the transactions, each updated value in the view is sent to the cache, which writes the changes through to the DB server.  (If the view's values are out-of-date with the DB, write will throw a StaleException.)
+
+Like the cache, the view is structured as a ("table:key" map) of (sorted value maps).  The "table:key" map has keys composed of the concatenation of (table identifier) and (key) strings; its values are the (sorted value maps) described next.  The (sorted value maps) are from (LONG\_MAX - read_timestamp) keys to values (JSON), as in the cache.  (Note, as mentioned above, the keys are (LONG\_MAX - read\_timestamp) rather than just (read\_timestamp) because we want the table's entries to be sorted in reverse, from most recent to oldest timestamp.)  
```

Comment:
There's no need for the second level. The entire view is read as-of the same read time. The cache maintains values read as-of at different read times, and keeps the historical values it finds as a result. The Tx view doesn't need to do that.